package/MDAnalysis/analysis/dssp.py

```diff
@@ -0,0 +1,279 @@
+    coord, org_shape = _check_input(coord)
+    b, l, a, _ = coord.shape
+    # add pseudo-H atom if not available
+    assert (a == 4) or (
+        a == 5
+    ), "Number of atoms should be 4 (N,CA,C,O) or 5 (N,CA,C,O,H)"
+    h = coord[:, 1:, 4] if a == 5 else _get_hydrogen_atom_position(coord)
+    # distance matrix
+    nmap = repeat(coord[:, 1:, 0], "... m c -> ... m n c", n=l - 1)
+    hmap = repeat(h, "... m c -> ... m n c", n=l - 1)
+    cmap = repeat(coord[:, 0:-1, 2], "... n c -> ... m n c", m=l - 1)
+    omap = repeat(coord[:, 0:-1, 3], "... n c -> ... m n c", m=l - 1)
+    d_on = np.linalg.norm(omap - nmap, axis=-1)
+    d_ch = np.linalg.norm(cmap - hmap, axis=-1)
+    d_oh = np.linalg.norm(omap - hmap, axis=-1)
+    d_cn = np.linalg.norm(cmap - nmap, axis=-1)
+    # electrostatic interaction energy
+    e = np.pad(
+        CONST_Q1Q2 * (1.0 / d_on + 1.0 / d_ch - 1.0 /
+                      d_oh - 1.0 / d_cn) * CONST_F,
+        [[0, 0], [1, 0], [0, 1]],
+    )
+    if return_e:
+        return e
+    # mask for local pairs (i,i), (i,i+1), (i,i+2)
+    local_mask = ~np.eye(l, dtype=bool)
+    local_mask *= ~np.diag(np.ones(l - 1, dtype=bool), k=-1)
+    local_mask *= ~np.diag(np.ones(l - 2, dtype=bool), k=-2)
+    # hydrogen bond map (continuous value extension of original definition)
+    hbond_map = np.clip(cutoff - margin - e, a_min=-margin, a_max=margin)
+    hbond_map = (np.sin(hbond_map / margin * np.pi / 2) + 1.0) / 2
+    hbond_map = hbond_map * repeat(local_mask, "l1 l2 -> b l1 l2", b=b)
+    # return h-bond map
+    hbond_map = np.squeeze(hbond_map, axis=0) if len(
+        org_shape) == 3 else hbond_map
+    return hbond_map
+
+
+def assign(coord: np.ndarray) -> np.ndarray:
+    """Assigns secondary structure for a given coordinate array
+
+    Args:
+        coord (np.ndarray): input coordinates in either (n, 4, 3) or (n, 5, 3) shape -- without or with hydrogens
+
+    Returns:
+        np.ndarray: output (n,) array with labels in C3 notation ('-', 'H', 'E')
+    """
+    # check input
+    coord, org_shape = _check_input(coord)
+    # get hydrogen bond map
+    hbmap = get_hbond_map(coord)
+    hbmap = rearrange(
+        hbmap, "... l1 l2 -> ... l2 l1"
+    )  # convert into "i:C=O, j:N-H" form
+    # identify turn 3, 4, 5
+    turn3 = np.diagonal(hbmap, axis1=-2, axis2=-1, offset=3) > 0.0
+    turn4 = np.diagonal(hbmap, axis1=-2, axis2=-1, offset=4) > 0.0
+    turn5 = np.diagonal(hbmap, axis1=-2, axis2=-1, offset=5) > 0.0
+    # assignment of helical sses
+    h3 = np.pad(turn3[:, :-1] * turn3[:, 1:], [[0, 0], [1, 3]])
+    h4 = np.pad(turn4[:, :-1] * turn4[:, 1:], [[0, 0], [1, 4]])
+    h5 = np.pad(turn5[:, :-1] * turn5[:, 1:], [[0, 0], [1, 5]])
+    # helix4 first
+    helix4 = h4 + np.roll(h4, 1, 1) + np.roll(h4, 2, 1) + np.roll(h4, 3, 1)
+    h3 = h3 * ~np.roll(helix4, -1, 1) * ~helix4  # helix4 is higher prioritized
+    h5 = h5 * ~np.roll(helix4, -1, 1) * ~helix4  # helix4 is higher prioritized
```

Comment:
Can we get 3_10 and π helix, too?

Reply:
I'm not certain, but seems like yes (used [this paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2981643/)'s Fig 1):

```python
from MDAnalysis.analysis.dssp import DSSP
import MDAnalysis as mda

u = mda.Universe('./1FUR.pdb')
r = DSSP(u).run()

r.results.resids[151:158]
# array([155, 156, 157, 158, 159, 160, 161])

''.join(r.results.dssp[0])[151:158]
# 'HHHH-HH'
```

Reply:
Yes, it can detect 3-10 and pi helixes indeed -- I generated idealized structures with `mmtbx` from [here](https://cci.lbl.gov/docs/cctbx/script_ideal_ss/) and ran DSSP on them:

```
dssp/o_beta_seq.pdb           --------------------
dssp/o_helix310_seq.pdb    -HHHHHHHHHHHHHHHHHH-
dssp/o_helix_seq.pdb          -HHHHHHHHHHHHHHHHHH-
o_helix_pi_seq.pdb              -HHHHHHHHHHHHHHHHHH-
```